lib/rex/proto/kerberos/client.rb

```diff
@@ -160,20 +162,41 @@ def send_request_udp(req)
         # @raise [Rex::Proto::Kerberos::Model::Error::KerberosDecodingError] if the response can't be processed
         # @raise [EOFError] if expected data can't be read
         def recv_response_tcp
-          length_raw = connection.get_once(4, timeout)
+          remaining = timeout
+          length_raw, elapsed_time = Rex::Stopwatch.elapsed_time do
+            connection.get_once(4, remaining)
+          end
+          remaining -= elapsed_time
           unless length_raw && length_raw.length == 4
-            raise ::EOFError, 'Kerberos Client: failed to read response'
+            if remaining <= 0
+              raise Rex::TimeoutError, 'Kerberos Client: failed to read response length due to timeout'
```

Comment:
I played around with the latency settings to force this timeout to be hit, but it looks like the user isn't notified about this error anymore as a result of changing the `::EOF` error to `Rex::TimeoutError` - I believe the Rex::Timeout is being swallowed by the exception handling higher up in the stack:

```
msf6 auxiliary(gather/ldap_query) > enum_accounts rhost=192.168.123.13 username=Administrator password=p4$$w0rd ldapauth=kerberos ldaprhostname=dc3.adf3.local domain=adf3.local domaincontrollerrhost=192.168.123.13
[*] Running module against 192.168.123.13

[-] Stopping running against current target...
[*] Control-C again to force quit all targets.
[*] Auxiliary module execution completed
```

Reply:
That makes sense but this should be a timeout error right? An EOF isn't reached and the socket is still open.

Reply:
A timeout error makes sense; I'm not sure the best way to ensure this gets logged out though 🤔 

At the minute, there's the potential of this timing out half way through an ldap query - and it's not that clear to the user what went wrong:

```
msf6 auxiliary(gather/ldap_query) > enum_accounts rhost=192.168.123.13 username=Administrator password=p4$$w0rd ldapauth=kerberos ldaprhostname=dc3.adf3.local domain=adf3.local domaincontrollerrhost=192.168.123.13 QUERY_ATTRIBUTES=a,b,c,d,e
[*] Running module against 192.168.123.13

[+] 192.168.123.13:88 - Received a valid TGT-Response
[*] 192.168.123.13:389 - TGT MIT Credential Cache ticket saved to /Users/user/.msf4/loot/20230110234313_default_192.168.123.13_mit.kerberos.cca_721205.bin
[+] 192.168.123.13:88 - Received a valid TGS-Response
[*] 192.168.123.13:389 - TGS MIT Credential Cache ticket saved to /Users/user/.msf4/loot/20230110234317_default_192.168.123.13_mit.kerberos.cca_962649.bin
[-] Stopping running against current target...
[*] Control-C again to force quit all targets.
[*] Auxiliary module execution completed
```

It also prints out a confusing message about interrupts:

```
[-] Stopping running against current target...
[*] Control-C again to force quit all targets.
[*] Auxiliary module execution completed
```

Reply:
I took a closer look at this - and it looks like `Rex::TimeoutError` is an `::Interrupt`, and we also lose the custom message that you've added:

```
From: /Users/user/Documents/code/metasploit-framework/lib/msf/base/simple/auxiliary.rb:199 Msf::Simple::Auxiliary.job_run_proc:

    194:       mod.print_error("Auxiliary triggered a timeout exception")
    195:       mod.cleanup
    196:       return
    197:     rescue ::Interrupt => e
    198:       require 'pry-byebug'; binding.pry
 => 199:       mod.error = e
    200:       mod.print_error("Stopping running against current target...")
    201:       mod.cleanup
    202:       mod.print_status("Control-C again to force quit all targets.")
    203:       begin
    204:         Rex.sleep(0.5)

[1] pry(Msf::Simple::Auxiliary)> e
=> #<Rex::TimeoutError: Operation timed out.>
[2] pry(Msf::Simple::Auxiliary)> e.is_a?(::Interrupt)
=> true
```

Current implementation:

https://github.com/rapid7/rex-core/blob/1e43b174888d673a045aead891ed9c4085ad85f7/lib/rex/exceptions.rb#L13-L24

A quick dig into the history shows the original implementation always extended `::Interrupt`, and it's not super clear to me why:

https://github.com/rapid7/metasploit-framework/commit/a2f34c4b734db2f10a96ee081008a9e2c216b4bd#diff-ffa8aecc620da4225c05d6cb16319a710fdf2f5b36c1053effe8897d28a914c8R16-R26

In comparison - Ruby's native timeout exception isn't an interrupt:

```
>> Timeout.timeout(1) { sleep 2 }
/Users/user/.rvm/gems/ruby-3.0.5@metasploit-framework/gems/rex-core-0.1.28/lib/rex/sync/thread_safe.rb:36:in `select': execution expired (Timeout::Error)
....
>> Timeout::Error.new.is_a?(::Interrupt)
=> false
>> 3.0.5 :002 > pp Timeout::Error.ancestors
[Timeout::Error,
 RuntimeError,                               
 StandardError,                              
 Exception,                                  
 Object,                                     
 PP::ObjectMixin,                            
 Kernel,                                     
 BasicObject]
```

I guess this is all an existing issue though; I'm happy to treat this as not a blocker though - since this PR fixes the more common scenario of hitting bandwidth issues

Reply:
Comment not a blocker; We'll get a separate PR for bumping the changes to rex-core into master https://github.com/rapid7/rex-core/pull/33, as well as making a change to https://github.com/rapid7/metasploit-framework/blob/61ffb9095489ef3c23e5b0d7f4618c6ba4a8120a/lib/msf/base/simple/auxiliary.rb#L192-L196 to add catching Rex::TimeoutError - which is required to separate it from the ::Interupt rescue which accidentally captures `Rex::TimeoutError` - https://github.com/rapid7/metasploit-framework/blob/61ffb9095489ef3c23e5b0d7f4618c6ba4a8120a/lib/msf/base/simple/auxiliary.rb#L197